### PR TITLE
More robust build process

### PR DESCRIPTION
**Pull request by ltratt**
This is intended to address the issues @smarr has pointed out in https://github.com/softdevteam/warmup_stats/issues/95. I have only lightly tested this so far, so wouldn't mind someone a bit more au fait with this part of the system giving it a once over before we consider merging it.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -87,10 +87,10 @@ echo "devtools::install_git('git://github.com/rkillick/changepoint', branch = 'm
 which lsb_release > /dev/null 2>&1
 if [ $? -eq 0 ]; then
     if [ "$(lsb_release -si)" = "Ubuntu" ]; then
-        pip install --target "${PIP_TARGET_DIR}" "rpy2==2.8.5" || exit $?
+        python2.7 -m pip install --target "${PIP_TARGET_DIR}" "rpy2==2.8.5" || exit $?
```
Comment (vext01) [inline build.sh]: Is that more portable than `pip2.7` then?

File: bin/diff_results [hunk 2]
```diff
@@ -48,9 +48,15 @@ import os
 import sys
 
 # R packages are stored relative to the top-level of the repo.
-if ('R_LIBS_USER' not in os.environ or 'rlibs' not in os.environ['R_LIBS_USER']):
-    os.environ['R_LIBS_USER'] = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
-                                             'work', 'rlibs')
+our_rlibs = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'work', 'rlibs')
```
Comment (vext01) [inline bin/diff_results]: Can any of this stuff be put in a function for re-usability?

File: build.sh (hunk 1 above)
Comment (smarr) [reply build.sh]: Likely, because on my system it's `pip-2.7`

**Comment by smarr** [comment]
This still doesn't solve my main complained, but yeah, sure.
I just don't understand the reason why you have your own way of dealing with dependencies. Python has its way, (well various ways, if you consider virtualenv), and R has its way.
You build on top of that your own system.

Why do you need this? Why do you insist on `work/pylibs`/`work/rlibs`? I can now use warmup_stats without code changes, but by simply creating dummy folders.

File: bin/diff_results (hunk 2 above)
Comment (ltratt) [reply bin/diff_results]: There are only two instances, so I'm not too worried.

**Comment by ltratt** [comment]
> This still doesn't solve my main complained, but yeah, sure.

What was your main complaint then?

> I just don't understand the reason why you have your own way of dealing with dependencies. 

Neither package manager is particularly high quality: pip, for example, has a tendency to spew rubbish over user's home directories or (worse) their system directories. Forcing things to be local means that we minimise the chances of damage to the user's system.

File: build.sh (hunk 1 above)
Comment (ltratt) [reply build.sh]: The binaries are (unhelpfully) called `pip2`, `pip3` and so on, without the minor version, on Debian. So it seems is much more portable to say `python2.7 -m pip ...`.

**Comment by ltratt** [comment]
> I can test this on Monday if you'd like.

I've now tested this on PDF plots and HTML tables, so I'm fairly hopeful it's OK.

File: build.sh (hunk 1 above)
Comment (vext01) [reply build.sh]: OK.

**Comment by vext01** [comment]
Hi guys,

I'm only joining the conversation mid-way, but a few comments based on what I read.
 
> Python has its way, (well various ways, if you consider virtualenv), and R has its way.
You build on top of that your own system.
> ...
> Why do you need this? Why do you insist on work/pylibs/work/rlibs

Stefan has a point here. But I can offer some background as to why we do (or did) things this way.

At the time we (mostly Sarah) wrote warmup_stats, the R shipped with Debian (our workhorse of choice) couldn't be used due to some package manager bug. So we had to build our own R, which is probably why the wrapper script exists in the first place.

[We had to also work around some setuptools bugs in our script -- again, probably fixed in packages by now]

We now use the system R, so it may well be the case that we can simplify the build system to use standard paths, but I'm not sure if it is a good use of our time. It doesn't greatly upset me that we install into a sub-directory, but I do acknowledge that it's less than ideal.

> pip, for example, has a tendency to spew rubbish over user's home directories.

I've only ever seen it write into `~/.local`.

> or (worse) their system directories.

That's the default on OpenBSD. On Debian (and probably other Linux?) `pip install` installs into `~/.local`, which I actually think is a better default. I don't see why you'd want to install into the system tree using pip. In fact, I don't recall a time I've ever needed to do that!

But I digress :)

**Comment by ltratt** [comment]
> It doesn't greatly upset me that we install into a sub-directory, but I do acknowledge that it's less than ideal.

I politely disagree :) I don't trust Pip as I've seen it go wrong in several weird ways (R's package manager I barely know, so I can't say anything good or bad about it). By installing a guaranteed local version, we isolate ourselves from a great deal of potential weirdness. This was an explicit design decision that was separate from the R system problems that we once had.

From an end user's perspective, this strikes me as unimportant: you run `build.sh` and things work. Like most software I build, I don't really know (or care) how it works, provided that it does work: it makes no difference to me if it uses some other Pip/R packages that were lurking around or not (the build time and disk space used are both trivial).

That said, this PR does make it more likely that "it works" is not an incorrect statement ;)

**Comment by vext01** [comment]
> I politely disagree :) I don't trust Pip as I've seen it go wrong in several weird ways

I think you misunderstood. I'm saying it doesn't greatly bother me (for this project, at least) that we install into a non-standard location (`work/...`). If I understand correctly, this is Stefan's gripe with our script.

> By installing a guaranteed local version, we isolate ourselves from a great deal of potential weirdness.

I don't understand this part though. We still use pip to install our stuff in our wrapper script. Our script just invokes pip overriding the install path. So if pip has weirdness, installing into a different location won't avoid it. Or have I missed something?

**Comment by ltratt** [comment]
> I don't understand this part though. We still use pip to install our stuff in our wrapper script. Our script just invokes pip overriding the install path. So if pip has weirdness, installing into a different location won't avoid it. Or have I missed something?

If Pip goes completely AWOL, there isn't anything we can do, clearly. But at least some of the weirdness was the interaction with existing installed packages. I didn't debug it deeply, because I simply wiped `~/.local` (more than once, over time). After a while, I gave up on that as a bad job. Installing into a guaranteed clean directory has never given us any problems.

**Comment by ltratt** [comment]
Anyway, we're perhaps getting into the theological aspects here. Even though it seems I misunderstood Stefan's complaint, I think this PR improves things anyway, so I think we can consider merging?

**Pull request merged**